File: scripts/chatanywhere-embeddings.js
// scripts/chatanywhere-embeddings.js
// Node script to call ChatAnywhere/OpenAI-compatible embeddings endpoint
// Usage: set env vars CHAT_ANYWHERE_API_KEY and CHAT_ANYWHERE_API_URL (optional) then run:
//    node scripts/chatanywhere-embeddings.js

(async function(){
  try {
    const apiKey = process.env.CHAT_ANYWHERE_API_KEY || process.env.OPENAI_API_KEY;
    if (!apiKey) {
      console.error('Missing CHAT_ANYWHERE_API_KEY or OPENAI_API_KEY in environment');
      process.exit(1);
    }

    const base = process.env.CHAT_ANYWHERE_API_URL || process.env.OPENAI_API_BASE || 'https://api.chatanywhere.tech/v1';
    const embedUrl = (base.endsWith('/v1') || base.endsWith('/v1/')) ? `${base.replace(/\/$/, '')}/embeddings` : `${base.replace(/\/$/, '')}/v1/embeddings`;

    const model = process.env.CHAT_ANYWHERE_EMBEDDING_MODEL || process.env.OPENAI_EMBEDDING_MODEL || 'text-embedding-3-small';

    const singleText = 'This is a test embedding from ChatAnywhere via Node.js';
    console.log('Embedding single text using', embedUrl, 'model', model);

    const res1 = await fetch(embedUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${apiKey}`
      },
      body: JSON.stringify({ input: singleText, model })
    });

    if (!res1.ok) {
      const t = await res1.text();
      throw new Error(`Embedding error: ${res1.status} ${t}`);
    }
    const j1 = await res1.json();
    const emb1 = j1?.data?.[0]?.embedding || j1?.embedding;
    console.log('Single embedding length:', emb1?.length || 0);

    const texts = ['First text', 'Second text', 'Third text'];
    const res2 = await fetch(embedUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${apiKey}`
      },
      body: JSON.stringify({ input: texts, model })
    });
    if (!res2.ok) {
      const t = await res2.text();
      throw new Error(`Batch embedding error: ${res2.status} ${t}`);
    }
    const j2 = await res2.json();
    const items = j2?.data || (Array.isArray(j2) ? j2 : null);
    if (!items) {
      console.warn('Unexpected batch response shape:', JSON.stringify(j2).slice(0,500));
    } else {
      for (let i=0;i<items.length;i++){
        console.log(`text ${i} embedding length:`, items[i]?.embedding?.length || 0);
      }
    }

    console.log('Done');
  } catch (err) {
    console.error('Error running embeddings script:', err);
    process.exit(1);
  }
})();

File: scripts/chatanywhere_embeddings.py
"""
chatanywhere_embeddings.py

Demonstrates creating embeddings against ChatAnywhere's OpenAI-compatible API endpoint
using the OpenAI Python client. This script can be used to test the embedding endpoint
and verify embedding vector lengths.

Usage:
  # set env vars (example)
  export OPENAI_API_KEY="sk-..." \
         OPENAI_API_BASE="https://api.chatanywhere.tech/v1"

  python scripts/chatanywhere_embeddings.py

You can also set `CHAT_ANYWHERE_API_KEY` and `CHAT_ANYWHERE_API_URL` instead of the
OpenAI-named env vars — the script will prefer `OPENAI_API_KEY`/`OPENAI_API_BASE` but
fall back to `CHAT_ANYWHERE_API_KEY`/`CHAT_ANYWHERE_API_URL`.

Install dependencies:
  pip install --upgrade openai

Note: Do NOT commit real API keys into source control. Use environment variables.
"""
from __future__ import annotations
import os
import sys
from typing import List

try:
    # New OpenAI Python client import pattern
    from openai import OpenAI
except Exception as e:
    print("Missing OpenAI Python client. Install with: pip install openai", file=sys.stderr)
    raise


def get_client() -> OpenAI:
    # Prefer the standard OpenAI env vars, but allow ChatAnywhere-specific names as fallback
    api_key = os.getenv("OPENAI_API_KEY") or os.getenv("CHAT_ANYWHERE_API_KEY")
    base = os.getenv("OPENAI_API_BASE") or os.getenv("CHAT_ANYWHERE_API_URL") or "https://api.chatanywhere.tech/v1"

    if not api_key:
        raise SystemExit(
            "Missing API key. Set OPENAI_API_KEY or CHAT_ANYWHERE_API_KEY in your environment."
        )

    # The OpenAI client accepts api_key and base_url parameters for service compatibility.
    client = OpenAI(api_key=api_key, base_url=base)
    return client


def embed_single_text(client: OpenAI, text: str, model: str = "text-embedding-3-small") -> List[float]:
    resp = client.embeddings.create(model=model, input=text)
    # Typical OpenAI-compatible response: resp.data[0].embedding
    emb = resp.data[0].embedding
    return emb


def embed_batch(client: OpenAI, texts: List[str], model: str = "text-embedding-3-small") -> List[List[float]]:
    resp = client.embeddings.create(model=model, input=texts)
    embeddings = [item.embedding for item in resp.data]
    return embeddings


def main():
    client = get_client()

    print("Creating single embedding for sample text...")
    emb = embed_single_text(client, "This is a test embedding from ChatAnywhere/OpenAI-compatible API")
    print("Embedding length:", len(emb))

    texts = ["First text", "Second text", "Third text"]
    print("Creating batch embeddings for", len(texts), "texts...")
    batch = embed_batch(client, texts)
    for i, e in enumerate(batch):
        print(f"text {i} embedding length: {len(e)}")


if __name__ == "__main__":
    main()
